### DISCUSSION TITLE
How to filter out ListObjects by a Pattern or containString

### DISCUSSION THREAD

**Discussion by hantsy** [aws/aws-sdk-kotlin#1203] [Q&A] [answered]
We want to list objects in a S3 bucket like this, 

```kotlin 
client
    .listObjects(
        ListObjectsRequest {
            bucket = bucketName
        }
    )
```

But the `ListObjectsRequest` only accepts a `prefix` to match the file name, not provide pattern(regex) or `containString` like to filter out the expected files.

Using the result `contents` to filter is bad performance.

```kotlin 
client
    .listObjects(
        ListObjectsRequest {
            bucket = bucketName
        }
    )
    .contents?.mapNotNull { it.key }?.filter { it.contains(...) }
```

Due to S3 Client `listObjects` can only  list 1000 files at most, we can not get all expected files in this approach.

BTW, how to do a pagination here, for example, count with same condition and query the result by page(or offset) and limit.

**Comment by lauzadis** [comment]
Hi, S3 does not support listing objects with a specific regex pattern, only prefix.

 To paginate, you can use `listObjectsV2` which has a `startAfter` request field to specify the last key from the previous request.

**Comment by aajtodd** [comment, marked as answer]
There is also a paginator for [listObjectsV2Paginated](https://sdk.amazonaws.com/kotlin/api/latest/s3/aws.sdk.kotlin.services.s3.paginators/list-objects-v2-paginated.html) that will handle the paging logic for you. See [pagination](https://docs.aws.amazon.com/sdk-for-kotlin/latest/developer-guide/pagination.html) in the developer guide for more info.
